text/0015-status-update.md

```diff
@@ -0,0 +1,397 @@
+mechanism, similar to a `RefCell`, where we would keep track at runtime of which regions
+were currently borrowed.
+
+This would work (although it would add significant complexity) in cases where we
+had foreknowledge of the position and size of an object, but I do not believe
+this could work when dealing with tables that host offsets, or at least not in
+the general case. The problem with offsets is that they are not verified ahead
+of time, and they often point to objects that do not have a known size. This
+means there is really no way to know exactly how much memory a table at a given
+offset is going to access. While it is possible to imagine complicated runtime
+schemes for tracking this, it would definitely require allocating additional
+data structures, and would significantly degrade the ergonomics of the API, and
+significantly increase the learning curve.
+
+I've been thinking about this problem quite a bit, and my opinion it is evidence
+of a fundamental tension in the design goals of this project.
+
+#### deciding on goals
+
+I believe that we currently have two conflicting goals. One of the rationale for
+this project is concern about the difficulty of getting developers up to speed
+on HarfBuzz, and how hard it is to add features there without accidentally
+introducing bugs. There is a hope that Rust's safety guarantees can give us a
+code base that is easier to modify and hack on; however there is *also* a desire
+to end up with code that is more-or-less equivalent to the existing C++.
+
+The basic tension is this: the ergonomics and safety advantages that Rust offers
+work by explicitly preventing certain behaviour, such as the aliasing of mutable
+variables. We cannot translate idiomatic C++ directly into Rust and get safety;
+we get safety by ruling out certain patterns.
+
+I think it is really worth stepping back and considering our goals. We
+would like to have a tool for authoring & manipulating font files, and we would
+like it to be a) fast and b) easy to use. This is something that we can achieve,
+but I do not think the result is going to look like it might look in C++. My
+feeling is that too much of our thinking (or at least my thinking!) so far has
+been focused on porting an existing design to Rust, and that this may be
+misguided.
+
+#### Subsetting is special
+
+A major focus of (and motivation for) this work has been hb-subset. I think this
+has influenced our thinking in important ways. I'm not entirely confident in my
+analysis here, so please correct any bad assumptions. With that said:
+
+- My sense is that the design of hb-subset is largely path-dependent on the
+  existing design of HarfBuzz. This is just an intuition. What this means to me,
+  though, is that if we were designing a subsetter from scratch, we might make
+  different choices.
+- Subsetting really feels like a special case. To the best of my understanding
+  it is used almost exclusively(?) by Google Fonts, and at a much larger scale.
```

Comment:
Thank you very much for your attempts and efforts.

Subsetting is also used extensively in China, and we have deployed an external subsetting service at [iconfont.cn](https://www.iconfont.cn/webfont?spm=a313x.7781069.1998910419.d81ec59f2#!/webfont/index). In addition, a large number of Alibaba's external ads use subsetting to reduce the font file size. For CJK fonts, subsetting is a must if you want to use them on the web.

We are desperately looking for a high-performance subsetting engine, and hb-subset provides us with a good direction and guidance. So for us, speed and correctness (without breaking the font) are both equally important.

Reply:
@yisibl fwiw we use hb-subset for prod web traffic, including CJK fonts, and have for some time now.

Reply:
@garretrieger @anthrotype can you enumerate some of the major uses of subsetting you are aware of outside Google Fonts, both by Google and otherwise? - for example, Chrome uses, Flutter uses, etc.

Reply:
@rsheeter Can you share a URL for googlefonts in which cases `hb-subset` is used?

Reply:
> @garretrieger @anthrotype can you enumerate some of the major uses of subsetting you are aware of outside Google Fonts, both by Google and otherwise? - for example, Chrome uses, Flutter uses, etc.

One thing I was confused about in our first conversation: we talked about subsetting being used in chrome when saving as a PDF, but it wasn't clear to me that this subsetting was using HarfBuzz? So I want to make sure we are explicitly focusing on users of hb-subset, and not just general applications of subsetting.

Reply:
@yisibl try something like https://fonts.googleapis.com/css2?family=Noto+Serif+JP&text=%E3%81%93%E3%82%93%E3%81%AB%E3%81%A1%E3%81%AF%E4%B8%96%E7%95%8C

Reply:
> explicitly focusing on users of hb-subset, and not just general applications of subsetting

hb-subset - and it's future Rust twin - are supposed to be general purpose so ... I think I'm struggling a little to understand the significance of the proposed distinction? 

We would also expect hb-subset to power https://www.w3.org/TR/IFT/.